packages/react-server/src/ReactFizzServer.js

```diff
@@ -97,119 +212,580 @@ function renderNode(request: Request, node: ReactNodeList): void {
     try {
       const result = type(props);
-      renderNode(result);
+      renderNode(request, parentBoundary, segment, result);
     } catch (x) {
       if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
         // Something suspended, we'll need to create a new segment and resolve it later.
-        request.pendingChunks++;
-        const newSegment = createSegment(request, node);
-        const ping = newSegment.ping;
+        const insertionIndex = segment.chunks.length;
+        const newSegment = createPendingSegment(request, insertionIndex, null);
+        const suspendedWork = createSuspendedWork(
+          request,
+          node,
+          parentBoundary,
+          newSegment,
+          null,
+        );
+        const ping = suspendedWork.ping;
         x.then(ping, ping);
         // TODO: Emit place holder
       } else {
-        reportError(request, x);
-        // TODO: Emit client render signal
+        // We can rethrow to terminate the rest of this tree.
+        throw x;
       }
     }
   } else if (typeof type === 'string') {
-    request.completedChunks.push(formatChunk(type, props));
+    // TODO: Split this to start/end and process children.
+    segment.chunks.push(formatChunk(type, props));
+  } else if (type === REACT_SUSPENSE_TYPE) {
+    // Each time we enter a suspense boundary, we split out into a new segment for
+    // the fallback so that we can later replace that segment with the content.
+    // This also lets us split out the main content even if it doesn't suspend,
+    // in case it ends up generating a large subtree of content.
+    const fallback: ReactNodeList = props.fallback;
+    const content: ReactNodeList = props.children;
+
+    const newBoundary = createSuspenseBoundary();
+
+    const insertionIndex = segment.chunks.length;
+    // The children of the boundary segment is actually the fallback.
+    const boundarySegment = createPendingSegment(
+      request,
+      insertionIndex,
+      newBoundary,
+    );
+    // We create suspended work for the fallback because we don't want to actually work
+    // on it yet in case we finish the main content, so we queue for later.
+    const suspendedFallbackWork = createSuspendedWork(
+      request,
+      fallback,
+      parentBoundary,
+      boundarySegment,
+      newBoundary.id, // This is the ID we want to give this fallback so we can replace it later.
+    );
+    // TODO: This should be queued at a separate lower priority queue so that we only work
+    // on preparing fallbacks if we don't have any more main content to work on.
+    request.pingedWork.push(suspendedFallbackWork);
+
+    // This segment is the actual child content. We can start rendering that immediately.
+    const contentRootSegment = createPendingSegment(request, 0, null);
+    // We mark the root segment as having its parent flushed. It's not really flushed but there is
+    // no parent segment so there's nothing to wait on.
```

Comment:
This is confusing. Could the property be named differently?

Reply:
This was one of the hardest part of model. I just figured it out two days ago. The modeling of the roots are pretty confusing regardless. Not sure it helps and might just make the nested case harder to get.

Reply:
If this was named after what _should_ happen instead of what _has_ happened, would that be something like `shouldFlushOnCompletion`(BecauseParentHasAlreadyFlushed)?

Not proposing a name change since I see what you mean that it could make other parts more confusing, just trying to understand.